executor/src/witgen/rows.rs

```diff
@@ -322,17 +323,39 @@ impl<'row, 'a, T: FieldElement> RowPair<'row, 'a, T> {
     ) -> Self {
         Self {
             current,
-            next,
+            next: Some(next),
             current_row_index,
             fixed_data,
             unknown_strategy,
         }
     }
 
+    /// Creates a new row pair from a single row, setting the next row to None.
+    pub fn from_single_row(
+        current: &'row Row<'a, T>,
+        current_row_index: DegreeType,
+        fixed_data: &'a FixedData<'a, T>,
+        unknown_strategy: UnknownStrategy,
+    ) -> Self {
+        Self {
+            current,
+            next: None,
+            current_row_index,
+            fixed_data,
+            unknown_strategy,
+        }
+    }
+
+    /// Gets the cell corresponding to the given polynomial reference.
+    ///
+    /// # Panics
+    /// Panics if the next row is accessed but the row pair has been constructed with
+    /// [RowPair::from_single_row].
     fn get_cell(&self, poly: &PolynomialReference) -> &Cell<T> {
-        match poly.next {
-            false => &self.current[&poly.poly_id()],
-            true => &self.next[&poly.poly_id()],
+        match (poly.next, self.next.as_ref()) {
```

Comment:
Can you check if this causes any performance loss?

Reply:
The benchmark is looking good!
```
Benchmarking keccak-executor-benchmark/keccak: Collecting 10 samples in estimated 383.39 s (10
keccak-executor-benchmark/keccak
                        time:   [35.825 s 36.410 s 36.976 s]
                        change: [-1.5126% +0.8904% +3.4923%] (p = 0.52 > 0.05)
                        No change in performance detected.
```